### PR TITLE
Feature: any / all filtering with tags

### PR DESCRIPTION
**Pull request by shamoon**
Migration of https://github.com/jonaswinkler/paperless-ng/pull/399

### CODE REVIEW

**Comment by shamoon** [comment]
Hello @paperless-ngx/frontend & @paperless-ngx/test This would be good to have folks test out before merging, it still works for me but its been a year! =)

**Comment by cleaverm** [comment]
@shamoon do you think this should get a `non-trivial` label as per #53?

**Comment by qcasey** [comment]
Jonas' last feedback on this issue:

> Also: Not sure how this will play out with nested tags. I'd say it's wise to postpone this PR until that's figured out.

Is that still relevant?

**Comment by shamoon** [comment]
Happy to call this non-trivial and have more folks take a look. 

And I do agree Jonas’ concern is worth considering but I’m not sure where it leaves us. Nested tags isn’t really on the radar yet so i suppose anything we do before that could add complexity if / when we do implement it. But unless someone has a suggestion for how to change this to avoid the possibility of adding any future complexity the only alternative is to just leave this out completely 🤷‍♂️ 

Up to the group.

File: src-ui/src/app/components/common/filterable-dropdown/filterable-dropdown.component.html [hunk 1]
```diff
@@ -12,22 +12,39 @@
...
           <use xlink:href="assets/bootstrap-icons.svg#arrow-right" />
         </svg>
       </button>
+      <div *ngIf="!editing" class="list-group-item list-group-item-note pt-1 pb-2">
+        <small i18n>Use
+          <svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" fill="currentColor" class="bi bi-option" viewBox="0 0 16 16">
+            <use xlink:href="assets/bootstrap-icons.svg#option" />
+          </svg>
+           + click / reutrn to exclude items.</small>
+      </div>
```
Comment (bauerj) [inline src-ui/src/app/components/common/filterable-dropdown/filterable-dropdown.component.html:47]: Small spelling issue

**Comment by bauerj** [comment]
@paperless-ngx/backend @paperless-ngx/frontend Please test and review.

**Comment by Nicarim** [comment]
I'll have a look at it now

**Comment by Nicarim** [comment]
[ ] Numbers on the tags dropdown are not updated once I change the tags in the documents - I have to refresh (But that doesn't happen in 1.5.0 too, so it's not related)
[ ] Choice of "Any" is not kept if I remove all selections and make another selection - not sure if it's intended, but from UX perspective I can imagine if you're working with "All" choice, you would want to do it several times, and having to change that each time might be annoying. It also works in contradiction with the rest of the interface (meaning, if you refresh with selected tags, that value is actually kept). 
[ ] How do you select NOT on linux? I can't seem to do it

**Comment by Nicarim** [comment]
I know why I can't use exclude on linux - the `alt` button is used as a shortcut to move a window in manjaro cinnamon - and firefox doesn't get the signal. I think I won't be the only one with this issue - so perhaps `ctrl` sounds better? or `shift`, not sure which of two is better.

**Comment by yschroeder** [comment]
Maybe we should find a way to exclude without needing a keyboard modifier? Our GUI might be used with touch input after all...

**Comment by Nicarim** [comment]
There could be a little "x" button on the right as well, yes. The touch screen input argument is really valid here, I can imagine paperless being ran on some tablet next to the scanner.

**Comment by shamoon** [comment]
> Numbers on the tags dropdown are not updated once I change the tags in the documents - I have to refresh (But that is a case in 1.5.0 too, so it's not related)
Yes, this is a current limitation of the app. I dont think its trivial to do but Jonas had said it wouldn't be too bad. Anyway, not part of this PR. 

> Choice of "Any" is not kept if I remove all selections and make another selection - not sure if it's intended, but from UX perspective I can imagine if you're working with "Any" choice, you would want to do it several times, and having to change that each time might be annoying. It also works in contradiction with the rest of the interface (meaning, if you refresh with selected tags, that value is actually kept).
Yes, reset is a full reset.

Great points (and in the chat) about the modifier key, I agree. So I think the best solution I think is to change it so that first click is ✔️, a second click is ❌. This also solves the mobile issue. A video is below.

If there are no other issues I think this is good to merge?

https://user-images.githubusercontent.com/4887959/154784758-d83ee1ed-6631-4389-a7b7-ba5748279283.mov

**Comment by yschroeder** [comment]
I tested this further: It works well with tags now. However, I have the feeling that is not thought through for correspondents and document types yet. I can do multiple selections on tags and exclude some of them. Why do correspondents and document types work differently? Why can't I select multiple correspondents to show up? Same for document types. Sure, every documents can only have one of those, while it can have multiple tags but still...

On the same note: The menus for correspondents and document types say in the bottom than I can click again to exclude, which is definitely wrong at this point.

TL;DR: Make correspondents and document type work like tags and I'm fine with this.

**Comment by shamoon** [comment]
Ah, well that is because the "has any correspondent" rule was not implemented in the backend. I agree that would be nice, if some was able to implement that I could update the UI. I agree having that note there is confusing so I removed that. This PR is just supposed to affect tags (see title), but I would love to enhance this feature in the [near] future.

**Comment by Nicarim** [comment]
I'm not sure if I'm convinced by double clicking to clear the filter, mainly because I feel like if you make too many selections you have to do much more work to remove it now, but it's personal preference tbh.  

It also resets the Any selection now if I want to unselect one item, and select another one, so it's even more frustrating now (before it was just if I remove all items that it resets like you said). Maybe it's value could be kept until you unselect every item?

File: src-ui/src/app/components/common/filterable-dropdown/filterable-dropdown.component.html [hunk 2]
```diff
@@ -12,22 +12,35 @@
...
       <button *ngIf="editing" class="list-group-item list-group-item-action bg-light" (click)="applyClicked()" [disabled]="!selectionModel.isDirty()">
-        <small class="ml-1" [ngClass]="{'font-weight-bold': selectionModel.isDirty()}" i18n>Apply</small>
+        <small class="ml-2" [ngClass]="{'font-weight-bold': selectionModel.isDirty()}" i18n>Apply</small>
         <svg width="1.5em" height="1em" viewBox="0 0 16 16" fill="currentColor">
           <use xlink:href="assets/bootstrap-icons.svg#arrow-right" />
         </svg>
       </button>
+      <div *ngIf="!editing && multiple" class="list-group-item list-group-item-note pt-1 pb-2">
+        <small i18n>Click again to exclude items.</small>
+      </div>
```
Comment (yschroeder) [inline src-ui/src/app/components/common/filterable-dropdown/filterable-dropdown.component.html:43]: This string needs to be added to the translation files:

https://angular.io/guide/i18n-common-translation-files

Comment (shamoon) [reply src-ui/src/app/components/common/filterable-dropdown/filterable-dropdown.component.html:43]: That happens via Crowdin automatically once merged into dev

**Comment by shamoon** [comment]
> I'm not sure if I'm convinced by double clicking to clear the filter, mainly because I feel like if you make too many selections you have to do much more work to remove it now, but it's personal preference tbh. But you can reset the filters on the right side, so that is not a problem I think.
> 
> It also resets the Any selection now if I want to unselect one item, and select another one, so it's even more frustrating now (before it was just if I remove all items that it resets like you said). Maybe it's value could be kept until you unselect every item?
> 
> Other than that I don't see any other issues.

Clicking just cycles through the states for each item Selected / Excluded / None, i.e. this is instead of using a modifier key. The reason any/all changes is because when you have at least 1 'excluded' item it automatically changes to 'all', that is what the API supports at the moment. I changed it to now make that more apparent to the user i.e. the button still stays visually selected.

**Comment by Nicarim** [comment]
Yes, I understand why it happens, my point was that now you have to go through this flow if you have 5 tags and want to check "Any" for 3 tags, and then switch to two tags:
1. Select Tag1, Tag2, Tag3
2. Select Any
3. Remove Tag2
4. You have to select Any again (since it resets)
There is no way now to remove a tag, without resetting the Any/All choice. It's just a small inconvenience, so honestly it depends on what others think.

**Comment by shamoon** [comment]
Ah, well yes an inconvenience but a trade-off of getting rid of the modifier key(s), there's no way to 'skip' the ❌ state. I think in terms of actual use cases its not as big a deal. Its not like people are going to be toggling 20 of these on-and-off all the time, they will likely be used for saved views and occasionally for looking for something specific. Yea, thats why I think its an acceptable inconvenience.

**Comment by yschroeder** [comment]
I also think the extra clicking is acceptable. LGTM.

**Review by yschroeder — APPROVED** [review]
